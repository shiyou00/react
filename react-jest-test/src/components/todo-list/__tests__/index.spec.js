import React from 'react';
import { shallow, render, mount} from 'enzyme';
import { TodoList } from "../index";

const props = {
  list: ['first', 'second'],
  deleteTodo: jest.fn(),
};

const setup = () => {
  const wrapper = shallow(<TodoList {...props} />);
  return {
    props,
    wrapper,
  };
};

const setupByRender = () => {
  const wrapper = render(<TodoList {...props} />);
  return {
    props,
    wrapper,
  };
};

const setupByMount = () => {
  const wrapper = mount(<TodoList {...props} />);
  return {
    props,
    wrapper,
  };
};

describe("todo-list", () => {

  it('renders correctly', () => {
    const wrapper = shallow(<TodoList {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should has Button', () => {
    const { wrapper } = setup();
    expect(wrapper.find('Button').length).toBe(2);
  });

  it('should render 2 item', () => {
    const { wrapper } = setupByRender();
    expect(wrapper.find('button').length).toBe(2);
  });

  it('should render item equal', () => {
    const { wrapper } = setupByMount();
    wrapper.find('.item-text').forEach((node, index) => {
      expect(node.text()).toBe(wrapper.props().list[index])
    });
  });

  it('click item to be done', () => {
    const { wrapper } = setupByMount();
    wrapper.find('Button').at(0).simulate('click');
    expect(props.deleteTodo).toBeCalled();
  });

  //使用spy替身的时候，在测试用例结束后，要对spy进行restore，不然这个spy会一直存在，并且无法对相同的方法再次进行spy。
  it('calls componentDidMount', () => {
    const componentDidMountSpy = jest.spyOn(TodoList.prototype, 'componentDidMount');
    const { wrapper } = setup();
    expect(componentDidMountSpy).toHaveBeenCalled();
    componentDidMountSpy.mockRestore();
  });

  it('calls component handleTest', () => { // class中使用箭头函数来定义方法
    const { wrapper } = setup();
    const spyFunction = jest.spyOn(wrapper.instance(), 'handleTest');
    wrapper.instance().handleTest();
    expect(spyFunction).toHaveBeenCalled();
    spyFunction.mockRestore();
  });

  it('calls component handleTest2', () => { //在constructor使用bind来定义方法
    const spyFunction = jest.spyOn(TodoList.prototype, 'handleTest2');
    const { wrapper } = setup();
    wrapper.instance().handleTest2();
    expect(spyFunction).toHaveBeenCalled();
    spyFunction.mockRestore();
  });

  it('test sum', () => {
    const { wrapper } = setup();
    expect(wrapper.instance().sum(2,3)).toEqual(5);
  });

});
