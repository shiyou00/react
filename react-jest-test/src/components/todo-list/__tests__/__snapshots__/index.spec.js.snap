// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todo-list renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TodoList
    deleteTodo={[MockFunction]}
    list={
      Array [
        "first",
        "second",
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div>
          <span
            className="item-text "
          >
            first
          </span>
          <Button
            block={false}
            ghost={false}
            htmlType="button"
            loading={false}
            onClick={[Function]}
          >
            done
          </Button>
        </div>,
        <div>
          <span
            className="item-text "
          >
            second
          </span>
          <Button
            block={false}
            ghost={false}
            htmlType="button"
            loading={false}
            onClick={[Function]}
          >
            done
          </Button>
        </div>,
      ],
      "className": "todo-list",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "0",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <span
              className="item-text "
            >
              first
            </span>,
            <Button
              block={false}
              ghost={false}
              htmlType="button"
              loading={false}
              onClick={[Function]}
            >
              done
            </Button>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "first",
              "className": "item-text ",
            },
            "ref": null,
            "rendered": "first",
            "type": "span",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "block": false,
              "children": "done",
              "ghost": false,
              "htmlType": "button",
              "loading": false,
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "done",
            "type": [Function],
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <span
              className="item-text "
            >
              second
            </span>,
            <Button
              block={false}
              ghost={false}
              htmlType="button"
              loading={false}
              onClick={[Function]}
            >
              done
            </Button>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "second",
              "className": "item-text ",
            },
            "ref": null,
            "rendered": "second",
            "type": "span",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "block": false,
              "children": "done",
              "ghost": false,
              "htmlType": "button",
              "loading": false,
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "done",
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div>
            <span
              className="item-text "
            >
              first
            </span>
            <Button
              block={false}
              ghost={false}
              htmlType="button"
              loading={false}
              onClick={[Function]}
            >
              done
            </Button>
          </div>,
          <div>
            <span
              className="item-text "
            >
              second
            </span>
            <Button
              block={false}
              ghost={false}
              htmlType="button"
              loading={false}
              onClick={[Function]}
            >
              done
            </Button>
          </div>,
        ],
        "className": "todo-list",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "0",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <span
                className="item-text "
              >
                first
              </span>,
              <Button
                block={false}
                ghost={false}
                htmlType="button"
                loading={false}
                onClick={[Function]}
              >
                done
              </Button>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "first",
                "className": "item-text ",
              },
              "ref": null,
              "rendered": "first",
              "type": "span",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "block": false,
                "children": "done",
                "ghost": false,
                "htmlType": "button",
                "loading": false,
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "done",
              "type": [Function],
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <span
                className="item-text "
              >
                second
              </span>,
              <Button
                block={false}
                ghost={false}
                htmlType="button"
                loading={false}
                onClick={[Function]}
              >
                done
              </Button>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "second",
                "className": "item-text ",
              },
              "ref": null,
              "rendered": "second",
              "type": "span",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "block": false,
                "children": "done",
                "ghost": false,
                "htmlType": "button",
                "loading": false,
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "done",
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Array [],
  Symbol(enzyme.__childContext__): null,
}
`;
